### PR TITLE
Add support for maps as data in changesets

### PR DESCRIPTION
**Pull request by zoldar**
Ecto supports passing `{data, types}` tuples for data, in addition to structs with schemas. Unfortunately, `phoenix_ecto` still assumes the data to be strictly a struct with a schema. This PR adds support for maps ~~and plain structs (without schema)~~ as data.

~~I'm not sure if https://github.com/phoenixframework/phoenix_ecto/pull/60/files#diff-5bbd15e070110d8f9b8d58d299755cf1R235 is the best way to tell apart a struct with schema from the one without but couldn't think of anything better at the moment.~~

**EDIT:** removed plain struct support - too questionable

### CODE REVIEW

File: lib/phoenix_ecto/html.ex [hunk 1]
```diff
@@ -231,17 +231,25 @@ if Code.ensure_loaded?(Phoenix.HTML) do
     defp form_for_errors(%{action: nil}), do: []
     defp form_for_errors(%{errors: errors}), do: errors
 
-    defp form_for_hidden(data) do
-      # Since they are primary keys, we should ignore nil values.
-      for {k, v} <- Ecto.primary_key(data), v != nil, do: {k, v}
+    defp form_for_hidden(%{__struct__: module} = data) do
+      if Keyword.has_key?(module.__info__(:functions), :__schema__) do
```
Comment (michalmuskala) [inline lib/phoenix_ecto/html.ex]: Should it just assume that if a struct is given it will be an ecto schema? I think it's fair to assume that.

Comment (zoldar) [inline lib/phoenix_ecto/html.ex]: Well, I'm not sure. Perhaps there can be some legitimate uses of plain structs too? (I personally needed raw maps but plain structs support didn't seem to be so costly to add, so I did it too)

Comment (michalmuskala) [inline lib/phoenix_ecto/html.ex]: If we're going to do this, we should use `function_exported?/3`. I'm not convinced, though.

Comment (zoldar) [inline lib/phoenix_ecto/html.ex]: Yeah, the `perhaps` part in my argumentation gives away how covincing it is :D

Okay then, I'll strip that thing off and leave only the catch-all match to accommodate for raw maps.

**Comment by zoldar** [comment]
I've removed support for plain structs - implementation was hacky, it's usefulness questionable.

File: lib/phoenix_ecto/html.ex [hunk 2]
```diff
@@ -231,17 +231,21 @@ if Code.ensure_loaded?(Phoenix.HTML) do
...
+    defp form_for_hidden(_), do: []
 
     defp form_for_name(%{__struct__: module}) do
       module
       |> Module.split()
       |> List.last()
       |> Macro.underscore()
     end
+    defp form_for_name(_) do
+      raise ArgumentError, "non-struct data passed in the changeset without `:as` option set"
```
Comment (josevalim) [inline lib/phoenix_ecto/html.ex]: This tell what went wrong but it doesn't tell what the user should do. Maybe we should rephrase it to something like "non-struct data in changeset requires the `:as` option to be given"?

**Comment by josevalim** [comment]
This looks great. I have added one tiny comment and I believe this is ready to go!
